presto-main/src/main/java/com/facebook/presto/operator/scalar/UnicodeUtil.java

```
@@ -0,0 +1,214 @@
+        while (i < string.length()) {
+            int i8 = string.getByte(i) & 0xff;
+            //
+            // Count bytes which are NOT the start of a code point
+            correctIndex += ((i8 >> 7) & (~i8 >> 6));
+            if (i == correctIndex) {
+                return i;
+            }
+
+            i++;
+        }
+
+        assert i == string.length();
+        return string.length();
+    }
+
+    /**
+     * Find substring within UTF-8 encoding string.
+     */
+    static int findUtf8IndexOfString(final Slice string, final int start, final int end, final Slice substring)
+    {
+        for (int i = start; i <= (end - substring.length()); i++) {
+            //
+            // TODO If slice provides indexOfByte with start, we could optimize searching
+            if (string.equals(i, substring.length(), substring, 0, substring.length())) {
+                return i;
+            }
+        }
+
+        return -1;
+    }
+
+    /**
+     * Method returns the length of code point by examine the start byte.
+     * <p>
+     * Invalid start bytes or code points will not throw an exception but return 1.
+     * </p>
+     */
+    static int requiredLengthOfCodePoint(final int ch)
+    {
+        if (ch < 0x80) {
+            // normal ASCII
+            return 1;
+        }
+        else if (ch < 0xc0) {
+            //
+            // 10xxxxxx -- illegal as start
+            // Illegal character but we want to skip this byte gracefully
+            return 1;
+        }
+        else if (ch < 0xe0) {
+            // 110xxxxx 10xxxxxx
+            return 2;
+        }
+        else if (ch < 0xf0) {
+            // 1110xxxx 10xxxxxx 10xxxxxx
+            return 3;
+        }
+        else if (ch < 0xf8) {
+            // 11110xxx 10xxxxxx 10xxxxxx 10xxxxxx
+            return 4;
+        }
+        //
+        // According to RFC3629 limited to 4 bytes so 5 and 6 bytes are illegal
+        return 1;
```

We should throw `PrestoException(INVALID_FUNCTION_ARGUMENT)` here. This means a connector provided corrupt data.